File: typography_document_model.py
"""Authoritative server-side mirror of the V20 TypographyDocumentModel.

The browser and server consume the same generated font registry contract. This
module normalizes the document-level semantic style catalog, object links, and
per-object overrides before the broader invitation validator runs.
"""
from __future__ import annotations
import copy,re,math
from typing import Any
from typography_contract import COLOR_TOKENS,MAX_TEXT_STYLES, finite_number, normalize_font_id, normalize_pairing_id, paired_font

MODEL_VERSION=1
KHMER_RE=re.compile(r"[\u1780-\u17ff\u19e0-\u19ff]")
STYLE_ID_RE=re.compile(r"^[a-z0-9][a-z0-9_-]{0,63}$")
HEX_RE=re.compile(r"^#[0-9a-fA-F]{6}$")
STYLE_FIELDS=("fontPairing","fontSize","textAutoFit","textAutoFitMax","textMinFontSize","fontWeight","fontStyle","lineHeight","letterSpacing","colorToken","color","textAlign","textVerticalAlign","textWrap","textColumns","textColumnGap","textPadding")
DEFAULT_STYLES={
 "display":{"id":"display","name":"Display","semantic":"display","fontPairing":"serif-formal","fontSize":64,"textAutoFit":"fit","textAutoFitMax":88,"textMinFontSize":18,"fontWeight":"700","fontStyle":"normal","lineHeight":1.08,"letterSpacing":0,"colorToken":"heading","textAlign":"center","textVerticalAlign":"middle","textWrap":"balance","textColumns":1,"textColumnGap":24,"textPadding":8,"builtin":True},
 "heading":{"id":"heading","name":"Heading","semantic":"heading","fontPairing":"serif-formal","fontSize":42,"textAutoFit":"fit","textAutoFitMax":56,"textMinFontSize":16,"fontWeight":"700","fontStyle":"normal","lineHeight":1.16,"letterSpacing":0,"colorToken":"heading","textAlign":"center","textVerticalAlign":"middle","textWrap":"balance","textColumns":1,"textColumnGap":24,"textPadding":8,"builtin":True},
 "subheading":{"id":"subheading","name":"Subheading","semantic":"subheading","fontPairing":"sans-modern","fontSize":28,"textAutoFit":"fit","textAutoFitMax":36,"textMinFontSize":14,"fontWeight":"700","fontStyle":"normal","lineHeight":1.25,"letterSpacing":0,"colorToken":"heading","textAlign":"center","textVerticalAlign":"middle","textWrap":"pretty","textColumns":1,"textColumnGap":24,"textPadding":8,"builtin":True},
 "body":{"id":"body","name":"Body","semantic":"body","fontPairing":"sans-modern","fontSize":18,"textAutoFit":"none","textAutoFitMax":22,"textMinFontSize":12,"fontWeight":"400","fontStyle":"normal","lineHeight":1.5,"letterSpacing":0,"colorToken":"text","textAlign":"left","textVerticalAlign":"top","textWrap":"pretty","textColumns":1,"textColumnGap":24,"textPadding":8,"builtin":True},
 "caption":{"id":"caption","name":"Caption","semantic":"caption","fontPairing":"sans-modern","fontSize":13,"textAutoFit":"none","textAutoFitMax":16,"textMinFontSize":10,"fontWeight":"400","fontStyle":"normal","lineHeight":1.4,"letterSpacing":.1,"colorToken":"muted","textAlign":"center","textVerticalAlign":"middle","textWrap":"normal","textColumns":1,"textColumnGap":16,"textPadding":6,"builtin":True},
 "khmer-ceremonial":{"id":"khmer-ceremonial","name":"Khmer Ceremonial","semantic":"khmer-ceremonial","fontPairing":"ceremonial-khmer","fontSize":48,"textAutoFit":"fit","textAutoFitMax":68,"textMinFontSize":18,"fontWeight":"700","fontStyle":"normal","lineHeight":1.42,"letterSpacing":0,"colorToken":"heading","textAlign":"center","textVerticalAlign":"middle","textWrap":"balance","textColumns":1,"textColumnGap":24,"textPadding":10,"builtin":True},
}
DEFAULT_ORDER=list(DEFAULT_STYLES)

def _number(value:Any,fallback:float,minimum:float,maximum:float,*,strict:bool,integer:bool=False):
 if strict and integer and isinstance(value,(int,float)) and not isinstance(value,bool) and math.isfinite(float(value)) and float(value)!=int(float(value)):raise ValueError("Integer typography value required")
 return finite_number(value,fallback,minimum,maximum,strict=strict,integer=integer)

def _clean_name(value:Any,fallback:str,*,strict:bool):
 if not isinstance(value,str):
  if strict:raise ValueError("Invalid text style name")
  return fallback
 value=re.sub(r"[\x00-\x1f\x7f]","",value).strip()
 if not value or len(value)>80:
  if strict:raise ValueError("Invalid text style name")
  return (value[:80] or fallback)
 return value

def _style_id(value:Any,*,strict:bool,fallback="body"):
 if isinstance(value,str) and STYLE_ID_RE.fullmatch(value):return value
 if strict:raise ValueError("Invalid text style ID")
 return fallback

def normalize_style(value:Any,style_id:str,*,strict=True):
 if not isinstance(value,dict):raise ValueError("Text style must be an object")
 semantic=str(value.get("semantic") or (style_id if style_id in DEFAULT_STYLES else "body"))[:40]
 base=DEFAULT_STYLES.get(semantic,DEFAULT_STYLES.get(style_id,DEFAULT_STYLES["body"]))
 font_size=_number(value.get("fontSize",base["fontSize"]),base["fontSize"],8,200,strict=strict)
 maximum=_number(value.get("textAutoFitMax",max(font_size,base["textAutoFitMax"])),max(font_size,base["textAutoFitMax"]),8,200,strict=strict)
 minimum=_number(value.get("textMinFontSize",base["textMinFontSize"]),base["textMinFontSize"],8,72,strict=strict)
 if minimum>maximum:
  if strict:raise ValueError("Minimum text size exceeds auto-fit maximum")
  minimum=maximum
 auto_fit=value.get("textAutoFit",base["textAutoFit"])
 if auto_fit not in {"none","fit"}:raise ValueError("Invalid text auto-fit mode")
 weight=str(value.get("fontWeight",base["fontWeight"]))
 if weight not in {"400","700"}:raise ValueError("Invalid font weight")
 font_style=value.get("fontStyle",base["fontStyle"])
 if font_style not in {"normal","italic"}:raise ValueError("Invalid font style")
 align=value.get("textAlign",base["textAlign"])
 if align not in {"left","center","right","justify"}:raise ValueError("Invalid text alignment")
 vertical=value.get("textVerticalAlign",base["textVerticalAlign"])
 if vertical not in {"top","middle","bottom"}:raise ValueError("Invalid vertical text alignment")
 wrap=value.get("textWrap",base["textWrap"])
 if wrap not in {"normal","balance","pretty"}:raise ValueError("Invalid text wrapping mode")
 token=value.get("colorToken",base["colorToken"])
 if token not in COLOR_TOKENS:raise ValueError("Invalid typography color token")
 result={
  "id":style_id,"name":_clean_name(value.get("name",base["name"]),base["name"],strict=strict),"semantic":semantic,
  "fontPairing":normalize_pairing_id(value.get("fontPairing",value.get("fontPairId",base["fontPairing"])),strict=strict),
  "fontSize":font_size,"textAutoFit":auto_fit,"textAutoFitMax":maximum,"textMinFontSize":minimum,"fontWeight":weight,"fontStyle":font_style,
  "lineHeight":_number(value.get("lineHeight",base["lineHeight"]),base["lineHeight"],.8,3,strict=strict),
  "letterSpacing":_number(value.get("letterSpacing",base["letterSpacing"]),base["letterSpacing"],-2,20,strict=strict),
  "colorToken":token,"textAlign":align,"textVerticalAlign":vertical,"textWrap":wrap,
  "textColumns":_number(value.get("textColumns",base["textColumns"]),base["textColumns"],1,3,strict=strict,integer=True),
  "textColumnGap":_number(value.get("textColumnGap",base["textColumnGap"]),base["textColumnGap"],0,64,strict=strict),
  "textPadding":_number(value.get("textPadding",base["textPadding"]),base["textPadding"],0,64,strict=strict),
  "builtin":bool(value.get("builtin",False)),
 }
 color=value.get("color")
 if color not in (None,""):
  if not isinstance(color,str) or not HEX_RE.fullmatch(color):raise ValueError("Invalid text style color")
  result["color"]=color.lower()
 return result

def normalize_catalog(value:Any,*,strict=True):
 if value in (None,{}):value={}
 if not isinstance(value,dict):raise ValueError("Typography catalog must be an object")
 raw_styles=value.get("styles",{})
 if raw_styles in (None,{}):raw_styles={}
 if not isinstance(raw_styles,dict) or len(raw_styles)>MAX_TEXT_STYLES:raise ValueError("Invalid typography style catalog")
 styles={}
 for style_id in DEFAULT_ORDER:
  merged={**copy.deepcopy(DEFAULT_STYLES[style_id]),**copy.deepcopy(raw_styles.get(style_id,{}) or {}),"id":style_id,"builtin":True}
  styles[style_id]=normalize_style(merged,style_id,strict=strict)
 for raw_id,raw_style in raw_styles.items():
  style_id=_style_id(raw_id,strict=strict)
  if style_id in styles:continue
  styles[style_id]=normalize_style(raw_style,style_id,strict=strict)
 raw_order=value.get("styleOrder",DEFAULT_ORDER)
 if not isinstance(raw_order,list) or len(raw_order)>MAX_TEXT_STYLES:raise ValueError("Invalid typography style order")
 order=[]
 for raw_id in raw_order:
  style_id=_style_id(raw_id,strict=strict)
  if style_id not in styles:raise ValueError("Typography style order references a missing style")
  if style_id not in order:order.append(style_id)
 for style_id in styles:
  if style_id not in order:order.append(style_id)
 default_id=_style_id(value.get("defaultStyleId","body"),strict=strict)
 if default_id not in styles:raise ValueError("Default typography style is missing")
 return {"version":MODEL_VERSION,"defaultStyleId":default_id,"styles":styles,"styleOrder":order}

def _pair_for_font(value:Any):
 font=normalize_font_id(value if value is not None else "noto-serif",strict=True)
 if font in {"noto-sans","noto-sans-khmer"}:return "sans-modern"
 if font=="sans-arial":return "modern-system"
 if font=="sans-trebuchet":return "friendly-system"
 if font=="serif-georgia":return "classic-system"
 return "serif-formal"

def infer_style_id(obj:dict,object_id=""):
 text=str(obj.get("html") or obj.get("text") or "")
 size=_number(obj.get("fontSize",32),32,8,200,strict=False)
 key=str(object_id).lower()
 if KHMER_RE.search(text) and size>=34:return "khmer-ceremonial"
 if key=="title" or size>=52:return "display"
 if size>=34:return "heading"
 if key=="subtitle" or size>=23:return "subheading"
 if size<=14:return "caption"
 return "body"

def _legacy_projection(obj:dict,*,strict=True):
 font_size=_number(obj.get("fontSize",32),32,8,200,strict=strict)
 maximum=_number(obj.get("textAutoFitMax",font_size),font_size,8,200,strict=strict)
 minimum=_number(obj.get("textMinFontSize",10),10,8,72,strict=strict)
 if minimum>maximum:raise ValueError("Minimum text size exceeds auto-fit maximum")
 raw_pair=obj.get("fontPairing",obj.get("fontPairId"))
 raw_token=obj.get("colorToken")
 return {
  "fontPairing":normalize_pairing_id(raw_pair,strict=strict) if raw_pair is not None else _pair_for_font(obj.get("font","noto-serif")),"fontSize":font_size,"textAutoFit":obj.get("textAutoFit","none"),"textAutoFitMax":maximum,"textMinFontSize":minimum,
  "fontWeight":str(obj.get("fontWeight","400")),"fontStyle":obj.get("fontStyle","normal"),
  "lineHeight":_number(obj.get("lineHeight",1.35),1.35,.8,3,strict=strict),"letterSpacing":_number(obj.get("letterSpacing",0),0,-2,20,strict=strict),
  "colorToken":raw_token if raw_token in COLOR_TOKENS else "text","color":str(obj.get("color")).lower() if HEX_RE.fullmatch(str(obj.get("color") or "")) else None,
  "textAlign":obj.get("textAlign","center"),"textVerticalAlign":obj.get("textVerticalAlign","middle"),"textWrap":obj.get("textWrap","normal"),
  "textColumns":_number(obj.get("textColumns",1),1,1,3,strict=strict,integer=True),"textColumnGap":_number(obj.get("textColumnGap",24),24,0,64,strict=strict),"textPadding":_number(obj.get("textPadding",8),8,0,64,strict=strict),
 }

def normalize_overrides(value:Any,*,strict=True):
 if value in (None,{}):return {}
 if not isinstance(value,dict):raise ValueError("Typography overrides must be an object")
 unknown=set(value)-set(STYLE_FIELDS)
 if unknown:raise ValueError("Unsupported typography override")
 probe=normalize_style({**DEFAULT_STYLES["body"],**value,"id":"probe","name":"Probe"},"probe",strict=strict)
 result={}
 for key in value:
  if key=="color":
   color=value[key]
   if not isinstance(color,str) or not HEX_RE.fullmatch(color):raise ValueError("Invalid typography override color")
   result[key]=color.lower()
  else:result[key]=probe[key]
 return result

def normalize_object_typography(obj:dict,object_id:str,catalog:dict,*,strict=True):
 if not isinstance(obj,dict):raise ValueError("Invalid design object")
 if obj.get("type") not in (None,"text","decoration"):return obj
 style_id=obj.get("textStyleId") or infer_style_id(obj,object_id)
 style_id=_style_id(style_id,strict=strict)
 if style_id not in catalog["styles"]:raise ValueError("Design object references a missing text style")
 base=catalog["styles"][style_id]
 overrides=normalize_overrides(obj.get("typographyOverrides"),strict=strict)
 legacy=_legacy_projection(obj,strict=strict)
 if not obj.get("typographyModelVersion"):
  for key in STYLE_FIELDS:
   if key in legacy and legacy[key] is not None and legacy[key]!=base.get(key):overrides.setdefault(key,legacy[key])
 elif obj.get("typographyResolvedSnapshot") is not None:
  snapshot=normalize_overrides(obj.get("typographyResolvedSnapshot"),strict=strict)
  for key in STYLE_FIELDS:
   if key in legacy and key in snapshot and legacy[key]!=snapshot[key]:overrides[key]=legacy[key]
 detached=obj.get("typographyDetached",False)
 if not isinstance(detached,bool):raise ValueError("Invalid typography detach state")
 resolved={**base,**(legacy if detached else overrides)}
 obj["typographyModelVersion"]=MODEL_VERSION;obj["textStyleId"]=style_id;obj["typographyDetached"]=detached;obj["typographyOverrides"]={k:v for k,v in overrides.items() if v is not None}
 locale="km" if KHMER_RE.search(str(obj.get("html") or "")) else "en"
 raw_font=obj.get("font")
 if raw_font is None:compatibility_font=paired_font(resolved["fontPairing"],"en")
 elif obj.get("typographyModelVersion"):
  compatibility_font=normalize_font_id(raw_font,strict=True)
 else:
  compatibility_font=normalize_font_id(raw_font,strict=True)
 obj["font"]=compatibility_font
 obj["fontPairing"]=resolved["fontPairing"]
 obj["colorToken"]=resolved.get("colorToken","text")
 for key in ("fontSize","textAutoFit","textAutoFitMax","textMinFontSize","fontWeight","fontStyle","lineHeight","letterSpacing","textAlign","textVerticalAlign","textWrap","textColumns","textColumnGap","textPadding"):
  obj[key]=resolved[key]
 if resolved.get("color"):obj["color"]=resolved["color"]
 obj["typographyResolvedSnapshot"]={key:resolved[key] for key in STYLE_FIELDS if key in resolved and resolved[key] is not None}
 return obj

def normalize_document_typography(document:dict,*,strict=True):
 catalog=normalize_catalog(document.get("typography"),strict=strict)
 document["typography"]=catalog
 def visit(mapping):
  if not isinstance(mapping,dict):raise ValueError("Invitation contains invalid design objects")
  for object_id,obj in mapping.items():normalize_object_typography(obj,object_id,catalog,strict=strict)
 visit(document.get("objects",{}))
 pages=document.get("designPages",[]) or []
 if not isinstance(pages,list):raise ValueError("Invalid visual pages")
 for page in pages:
  if isinstance(page,dict):visit(page.get("objects",{}))
 return document

File: src/js/typography-document-model.js
(()=>{'use strict';
const MODEL_VERSION=1,MAX_TEXT_STYLES=globalThis.EInviteTypography?.MAX_TEXT_STYLES||64;
const KHMER_RE=/[\u1780-\u17ff\u19e0-\u19ff]/u;
const STYLE_FIELDS=Object.freeze(['fontPairing','fontSize','textAutoFit','textAutoFitMax','textMinFontSize','fontWeight','fontStyle','lineHeight','letterSpacing','colorToken','color','textAlign','textVerticalAlign','textWrap','textColumns','textColumnGap','textPadding']);
const OVERRIDE_FIELDS=new Set(STYLE_FIELDS);
const clone=value=>typeof structuredClone==='function'?structuredClone(value):JSON.parse(JSON.stringify(value));
const own=(o,k)=>Object.prototype.hasOwnProperty.call(o,k);
const finite=(value,fallback,min,max)=>globalThis.EInviteTypography?.finiteNumber?EInviteTypography.finiteNumber(value,fallback,min,max):Number.isFinite(Number(value))?Math.max(min,Math.min(max,Number(value))):fallback;
const fontId=value=>globalThis.EInviteFontRegistry?.fontId?EInviteFontRegistry.fontId(value):'noto-serif';
const pairingId=value=>globalThis.EInviteFontRegistry?.pairingId?EInviteFontRegistry.pairingId(value):'serif-formal';
const pairForFont=value=>{const id=fontId(value);if(id==='noto-sans'||id==='noto-sans-khmer')return'sans-modern';if(id==='sans-arial')return'modern-system';if(id==='sans-trebuchet')return'friendly-system';if(id==='serif-georgia')return'classic-system';return'serif-formal'};
const safeId=(value,fallback='style')=>{const raw=String(value||'').trim().toLowerCase().replace(/[^a-z0-9_-]+/g,'-').replace(/^-+|-+$/g,'').slice(0,64);return raw||fallback};
const cleanName=(value,fallback)=>String(value||fallback).trim().replace(/[\u0000-\u001f\u007f]/g,'').slice(0,80)||fallback;
const DEFAULT_STYLES=Object.freeze({
 'display':Object.freeze({id:'display',name:'Display',semantic:'display',fontPairing:'serif-formal',fontSize:64,textAutoFit:'fit',textAutoFitMax:88,textMinFontSize:18,fontWeight:'700',fontStyle:'normal',lineHeight:1.08,letterSpacing:0,colorToken:'heading',textAlign:'center',textVerticalAlign:'middle',textWrap:'balance',textColumns:1,textColumnGap:24,textPadding:8,builtin:true}),
 'heading':Object.freeze({id:'heading',name:'Heading',semantic:'heading',fontPairing:'serif-formal',fontSize:42,textAutoFit:'fit',textAutoFitMax:56,textMinFontSize:16,fontWeight:'700',fontStyle:'normal',lineHeight:1.16,letterSpacing:0,colorToken:'heading',textAlign:'center',textVerticalAlign:'middle',textWrap:'balance',textColumns:1,textColumnGap:24,textPadding:8,builtin:true}),
 'subheading':Object.freeze({id:'subheading',name:'Subheading',semantic:'subheading',fontPairing:'sans-modern',fontSize:28,textAutoFit:'fit',textAutoFitMax:36,textMinFontSize:14,fontWeight:'700',fontStyle:'normal',lineHeight:1.25,letterSpacing:0,colorToken:'heading',textAlign:'center',textVerticalAlign:'middle',textWrap:'pretty',textColumns:1,textColumnGap:24,textPadding:8,builtin:true}),
 'body':Object.freeze({id:'body',name:'Body',semantic:'body',fontPairing:'sans-modern',fontSize:18,textAutoFit:'none',textAutoFitMax:22,textMinFontSize:12,fontWeight:'400',fontStyle:'normal',lineHeight:1.5,letterSpacing:0,colorToken:'text',textAlign:'left',textVerticalAlign:'top',textWrap:'pretty',textColumns:1,textColumnGap:24,textPadding:8,builtin:true}),
 'caption':Object.freeze({id:'caption',name:'Caption',semantic:'caption',fontPairing:'sans-modern',fontSize:13,textAutoFit:'none',textAutoFitMax:16,textMinFontSize:10,fontWeight:'400',fontStyle:'normal',lineHeight:1.4,letterSpacing:.1,colorToken:'muted',textAlign:'center',textVerticalAlign:'middle',textWrap:'normal',textColumns:1,textColumnGap:16,textPadding:6,builtin:true}),
 'khmer-ceremonial':Object.freeze({id:'khmer-ceremonial',name:'Khmer Ceremonial',semantic:'khmer-ceremonial',fontPairing:'ceremonial-khmer',fontSize:48,textAutoFit:'fit',textAutoFitMax:68,textMinFontSize:18,fontWeight:'700',fontStyle:'normal',lineHeight:1.42,letterSpacing:0,colorToken:'heading',textAlign:'center',textVerticalAlign:'middle',textWrap:'balance',textColumns:1,textColumnGap:24,textPadding:10,builtin:true})
});
const DEFAULT_ORDER=Object.freeze(Object.keys(DEFAULT_STYLES));
function normalizeStyle(input={},idHint='style'){
 const id=safeId(input.id||idHint,idHint),base=DEFAULT_STYLES[input.semantic]||DEFAULT_STYLES.body;
 const fontSize=finite(input.fontSize,base.fontSize,8,200),max=finite(input.textAutoFitMax,Math.max(fontSize,base.textAutoFitMax),8,200),min=Math.min(finite(input.textMinFontSize,base.textMinFontSize,8,72),max);
 const style={id,name:cleanName(input.name,base.name),semantic:String(input.semantic||base.semantic||'custom').slice(0,40),fontPairing:pairingId(input.fontPairing||input.fontPairId||base.fontPairing),fontSize,textAutoFit:input.textAutoFit==='fit'?'fit':'none',textAutoFitMax:max,textMinFontSize:min,fontWeight:String(input.fontWeight)==='700'?'700':'400',fontStyle:input.fontStyle==='italic'?'italic':'normal',lineHeight:finite(input.lineHeight,base.lineHeight,.8,3),letterSpacing:finite(input.letterSpacing,base.letterSpacing,-2,20),colorToken:(globalThis.EInviteTypography?.data?.colorTokens||['heading','text','accent','surface','muted','inverse']).includes(input.colorToken)?input.colorToken:base.colorToken,textAlign:['left','center','right','justify'].includes(input.textAlign)?input.textAlign:base.textAlign,textVerticalAlign:['top','middle','bottom'].includes(input.textVerticalAlign)?input.textVerticalAlign:base.textVerticalAlign,textWrap:['normal','balance','pretty'].includes(input.textWrap)?input.textWrap:base.textWrap,textColumns:Math.round(finite(input.textColumns,base.textColumns,1,3)),textColumnGap:finite(input.textColumnGap,base.textColumnGap,0,64),textPadding:finite(input.textPadding,base.textPadding,0,64),builtin:input.builtin===true};
 if(/^#[0-9a-f]{6}$/i.test(String(input.color||'')))style.color=String(input.color).toLowerCase();
 return style;
}
function defaultCatalog(){return{version:MODEL_VERSION,defaultStyleId:'body',styles:Object.fromEntries(DEFAULT_ORDER.map(id=>[id,clone(DEFAULT_STYLES[id])])),styleOrder:[...DEFAULT_ORDER]}}
function inferStyleId(object={},objectId=''){
 const text=String(object.html||object.text||''),size=finite(object.fontSize,32,8,200),id=String(objectId||object.id||'').toLowerCase();
 if(KHMER_RE.test(text)&&size>=34)return'khmer-ceremonial';if(id==='title'||size>=52)return'display';if(size>=34)return'heading';if(id==='subtitle'||size>=23)return'subheading';if(size<=14)return'caption';return'body';
}
function legacyProjection(object={}){
 const fontSize=finite(object.fontSize,32,8,200),max=finite(object.textAutoFitMax,fontSize,8,200);
 const rawPair=object.fontPairing||object.fontPairId;
 const rawToken=object.colorToken;
 return{fontPairing:rawPair?pairingId(rawPair):pairForFont(object.font),fontSize,textAutoFit:object.textAutoFit==='fit'?'fit':'none',textAutoFitMax:max,textMinFontSize:Math.min(finite(object.textMinFontSize,10,8,72),max),fontWeight:String(object.fontWeight)==='700'?'700':'400',fontStyle:object.fontStyle==='italic'?'italic':'normal',lineHeight:finite(object.lineHeight,1.35,.8,3),letterSpacing:finite(object.letterSpacing,0,-2,20),colorToken:(globalThis.EInviteTypography?.data?.colorTokens||['heading','text','accent','surface','muted','inverse']).includes(rawToken)?rawToken:'text',color:/^#[0-9a-f]{6}$/i.test(String(object.color||''))?String(object.color).toLowerCase():undefined,textAlign:['left','center','right','justify'].includes(object.textAlign)?object.textAlign:'center',textVerticalAlign:['top','middle','bottom'].includes(object.textVerticalAlign)?object.textVerticalAlign:'middle',textWrap:['normal','balance','pretty'].includes(object.textWrap)?object.textWrap:'normal',textColumns:Math.round(finite(object.textColumns,1,1,3)),textColumnGap:finite(object.textColumnGap,24,0,64),textPadding:finite(object.textPadding,8,0,64)};
}
function normalizeOverrides(input={}){const out={};if(!input||typeof input!=='object'||Array.isArray(input))return out;for(const key of STYLE_FIELDS)if(own(input,key)){const probe=normalizeStyle({[key]:input[key],semantic:'body'},'probe');if(key==='color'){if(/^#[0-9a-f]{6}$/i.test(String(input[key])))out[key]=String(input[key]).toLowerCase()}else out[key]=probe[key]}return out}
function differences(source,base){const out={};for(const key of STYLE_FIELDS){const a=source[key],b=base[key];if(a!==undefined&&JSON.stringify(a)!==JSON.stringify(b))out[key]=a}return out}
function normalizeObject(objectId,input={},catalog=defaultCatalog()){
 const o={...input};if(!['text','decoration'].includes(o.type||'text'))return o;const compatibilityFont=fontId(o.font);
 let styleId=safeId(o.textStyleId||inferStyleId(o,objectId),'body');if(!catalog.styles[styleId])styleId=catalog.defaultStyleId||'body';
 const base=catalog.styles[styleId]||catalog.styles.body||normalizeStyle(DEFAULT_STYLES.body,'body');
 let overrides=normalizeOverrides(o.typographyOverrides);const legacy=legacyProjection(o);
 if(!o.typographyModelVersion){overrides={...differences(legacy,base),...overrides}}
 else if(o.typographyResolvedSnapshot&&typeof o.typographyResolvedSnapshot==='object'&&!Array.isArray(o.typographyResolvedSnapshot)){
  const snapshot=normalizeOverrides(o.typographyResolvedSnapshot);
  for(const key of STYLE_FIELDS)if(legacy[key]!==undefined&&snapshot[key]!==undefined&&JSON.stringify(legacy[key])!==JSON.stringify(snapshot[key]))overrides[key]=legacy[key];
 }
 o.typographyModelVersion=MODEL_VERSION;o.textStyleId=styleId;o.typographyDetached=o.typographyDetached===true;o.typographyOverrides=overrides;
 const resolved=resolveObjectTypography({typography:catalog,palette:{}},o,{text:o.html||'',ignoreDetached:false});
 o.font=compatibilityFont;o.fontPairing=resolved.fontPairing;o.colorToken=resolved.colorToken;
 for(const key of ['fontSize','textAutoFit','textAutoFitMax','textMinFontSize','fontWeight','fontStyle','lineHeight','letterSpacing','textAlign','textVerticalAlign','textWrap','textColumns','textColumnGap','textPadding'])o[key]=resolved[key];
 if(resolved.color)o.color=resolved.color;
 o.typographyResolvedSnapshot=Object.fromEntries(STYLE_FIELDS.filter(key=>resolved[key]!==undefined).map(key=>[key,resolved[key]]));
 return o;
}
function normalizeCatalog(input={}){
 const defaults=defaultCatalog(),rawStyles=input&&typeof input==='object'&&input.styles&&typeof input.styles==='object'&&!Array.isArray(input.styles)?input.styles:{};const styles={};
 for(const id of DEFAULT_ORDER)styles[id]=normalizeStyle({...DEFAULT_STYLES[id],...(rawStyles[id]||{}),id},id);
 for(const [id,value] of Object.entries(rawStyles).slice(0,MAX_TEXT_STYLES)){const clean=safeId(id);if(!styles[clean])styles[clean]=normalizeStyle({...value,id:clean},clean)}
 const order=[];for(const id of Array.isArray(input.styleOrder)?input.styleOrder:DEFAULT_ORDER){const clean=safeId(id);if(styles[clean]&&!order.includes(clean))order.push(clean)}for(const id of Object.keys(styles))if(!order.includes(id))order.push(id);
 const defaultStyleId=styles[safeId(input.defaultStyleId||'body')]?safeId(input.defaultStyleId||'body'):'body';return{version:MODEL_VERSION,defaultStyleId,styles,styleOrder:order.slice(0,MAX_TEXT_STYLES)};
}
function normalizeDocument(input={},options={}){
 const doc=options.mutate?input:clone(input||{});if(globalThis.EInviteCustomFonts)EInviteCustomFonts.normalizeDocumentFonts(doc);doc.typography=normalizeCatalog(doc.typography||{});
 const normalizeMap=map=>{if(!map||typeof map!=='object'||Array.isArray(map))return{};for(const [id,obj] of Object.entries(map))if(obj&&typeof obj==='object')map[id]=normalizeObject(id,obj,doc.typography);return map};
 doc.objects=normalizeMap(doc.objects||{});for(const page of Array.isArray(doc.designPages)?doc.designPages:[])page.objects=normalizeMap(page.objects||{});return doc;
}
function objectText(object={},fallback=''){return String(object.html||object.text||object.content||fallback).replace(/<[^>]+>/g,' ')}
function detectLocale(text='',explicit='auto'){if(String(explicit).toLowerCase().startsWith('km'))return'km';if(String(explicit).toLowerCase().startsWith('en'))return'en';return KHMER_RE.test(String(text))?'km':'en'}
function resolveColorToken(document={},token='text'){const palette=document.palette||{},accent=document.accent||'#9d4555',map={heading:palette.heading||accent,text:palette.text||'#342c26',accent,surface:palette.surface||'#ffffff',muted:palette.muted||'#6f6770',inverse:'#ffffff'};return map[token]||map.text}
function resolveObjectTypography(document={},object={},options={}){
 const catalog=normalizeCatalog(document.typography||{}),styleId=catalog.styles[object.textStyleId]?object.textStyleId:catalog.defaultStyleId,style=catalog.styles[styleId]||catalog.styles.body;
 const legacy=legacyProjection(object),detached=object.typographyDetached===true,overrides=normalizeOverrides(object.typographyOverrides);let result=detached?{...style,...legacy}:{...style,...overrides};
 const text=options.text??objectText(object),locale=detectLocale(text,options.locale||object.textLocale||'auto'),font=globalThis.EInviteFontRegistry?.pairedFont?EInviteFontRegistry.pairedFont(result.fontPairing,locale):fontId(object.font),fontStack=globalThis.EInviteFontRegistry?.stack?EInviteFontRegistry.stack(font):'serif';
 const max=finite(result.textAutoFitMax,result.fontSize,8,200);result={...result,id:styleId,styleId,styleName:style.name,locale,font,compatibilityFont:fontId(object.font||font),fontStack,fontSize:finite(result.fontSize,style.fontSize,8,200),textAutoFit:result.textAutoFit==='fit'?'fit':'none',textAutoFitMax:max,textMinFontSize:Math.min(finite(result.textMinFontSize,style.textMinFontSize,8,72),max),fontWeight:String(result.fontWeight)==='700'?'700':'400',fontStyle:result.fontStyle==='italic'?'italic':'normal',lineHeight:finite(result.lineHeight,style.lineHeight,.8,3),letterSpacing:finite(result.letterSpacing,style.letterSpacing,-2,20),textAlign:['left','center','right','justify'].includes(result.textAlign)?result.textAlign:style.textAlign,textVerticalAlign:['top','middle','bottom'].includes(result.textVerticalAlign)?result.textVerticalAlign:style.textVerticalAlign,textWrap:['normal','balance','pretty'].includes(result.textWrap)?result.textWrap:style.textWrap,textColumns:Math.round(finite(result.textColumns,style.textColumns,1,3)),textColumnGap:finite(result.textColumnGap,style.textColumnGap,0,64),textPadding:finite(result.textPadding,style.textPadding,0,64)};
 result.color=/^#[0-9a-f]{6}$/i.test(String(result.color||''))?result.color:resolveColorToken(document,result.colorToken);return result;
}
function walkTextObjects(document,callback){const visit=map=>Object.entries(map||{}).forEach(([id,o])=>{if(o&&['text','decoration'].includes(o.type||'text'))callback(o,id,map)});visit(document.objects);for(const page of document.designPages||[])visit(page.objects)}
function uniqueStyleId(catalog,seed='custom-style'){const base=safeId(seed,'custom-style');let id=base,n=2;while(catalog.styles[id])id=`${base}-${n++}`;return id}
function createStyle(document,input={}){document.typography=normalizeCatalog(document.typography||{});if(Object.keys(document.typography.styles).length>=MAX_TEXT_STYLES)throw new RangeError(`Text style limit (${MAX_TEXT_STYLES}) reached`);const id=uniqueStyleId(document.typography,input.id||input.name||'custom-style'),style=normalizeStyle({...input,id,builtin:false},id);document.typography.styles[id]=style;document.typography.styleOrder.push(id);return id}
function renameStyle(document,id,name){document.typography=normalizeCatalog(document.typography||{});if(!document.typography.styles[id])throw new Error('Text style not found');document.typography.styles[id].name=cleanName(name,document.typography.styles[id].name);return id}
function duplicateStyle(document,id,name){document.typography=normalizeCatalog(document.typography||{});const source=document.typography.styles[id];if(!source)throw new Error('Text style not found');return createStyle(document,{...clone(source),id:'',name:name||`${source.name} Copy`,builtin:false})}
function updateStyle(document,id,patch={}){document.typography=normalizeCatalog(document.typography||{});const source=document.typography.styles[id];if(!source)throw new Error('Text style not found');document.typography.styles[id]=normalizeStyle({...source,...patch,id,builtin:source.builtin===true},id);return document.typography.styles[id]}
function deleteStyle(document,id,replacementId='body'){document.typography=normalizeCatalog(document.typography||{});if(!document.typography.styles[id])throw new Error('Text style not found');if(id===replacementId)throw new Error('Replacement style must be different');if(!document.typography.styles[replacementId])throw new Error('Replacement style not found');walkTextObjects(document,o=>{if(o.textStyleId===id)o.textStyleId=replacementId});delete document.typography.styles[id];document.typography.styleOrder=document.typography.styleOrder.filter(x=>x!==id);if(document.typography.defaultStyleId===id)document.typography.defaultStyleId=replacementId;return replacementId}
function linkObject(object,styleId){object.textStyleId=styleId;object.typographyDetached=false;object.typographyOverrides={};object.typographyModelVersion=MODEL_VERSION;return object}
function detachObject(document,object){const resolved=resolveObjectTypography(document,object);object.typographyDetached=true;object.typographyOverrides=Object.fromEntries(STYLE_FIELDS.filter(k=>resolved[k]!==undefined).map(k=>[k,resolved[k]]));return object}
function setOverride(object,key,value){if(!OVERRIDE_FIELDS.has(key))throw new Error('Unsupported typography override');object.typographyOverrides=normalizeOverrides({...object.typographyOverrides,[key]:value});object.typographyDetached=false;return object}
function resetOverride(object,key){if(!object.typographyOverrides)return object;if(key)delete object.typographyOverrides[key];else object.typographyOverrides={};return object}
function linkedCount(document,styleId){let count=0;walkTextObjects(document,o=>{if(o.textStyleId===styleId&&!o.typographyDetached)count++});return count}
window.TypographyDocumentModel=Object.freeze({MODEL_VERSION,MAX_TEXT_STYLES,STYLE_FIELDS,DEFAULT_STYLES,defaultCatalog,normalizeStyle,normalizeCatalog,normalizeObject,normalizeDocument,resolveObjectTypography,resolveColorToken,detectLocale,walkTextObjects,createStyle,renameStyle,duplicateStyle,updateStyle,deleteStyle,linkObject,detachObject,setOverride,resetOverride,linkedCount,clone});
})();
